### DISCUSSION TITLE
Head Mouse Prototype

### DISCUSSION THREAD

**Discussion by gbishop** [UNC-Project-Open-AAC/OS-DPI#211] [Announcements]
This is truly a hack. I started with [tracky-mouse](https://github.com/1j01/tracky-mouse) and did the minimum to get it working in our app. It **really** needs to be rewritten but seems to work. 

I have not yet pushed it to production but you can try it from my repo.

https://gbishop.github.io/OS-DPI/?fetch=examples/updated/headmouse_example.osdpi

I configured the example with keyboard controls (see the KeyHandler in the Pointer dwell method).

- **t** starts tracking and will ask for permission to use your camera the first time.
- **T** stops tracking and stops using the camera.
- **s** should open the tracky-mouse controls to let you see the camera and adjust several parameters. 

I don't hide the normal mouse pointer but things get weird if you use both simultaneously because things aren't internally designed for multiple mice.

There is a single state variable that defaults to `$HeadMouse`. Set it to "on" to start tracking, "off" to stop tracking or "setup" to open the controls. See the Actions with `Pointer dwell` as their origin. 

Please give it a try and let me know if it works for you. I have been able to operate it from greater than 6 feet away by turning up the horizontal and vertical sensitivity.

**Comment by aetkie** [comment]
This is cool Gary!

Head mouse works for me to interact with the design. Keyboard controls were not working for me, but I will keep playing around a bit.

**Reply by gbishop** [reply]
Please tell me about what is not working.

**Reply by aetkie** [reply]
If I understand your previous post correctly, the keyboard controls do not work within that design. 

Specifically, I type "T" and it does not stop tracking. 

Hitting the "s" key does not open the tracky-mouse controls 

Once I allowed my camera access, the head mouse begins tracking as soon as I open the UI. So I don't have to use the keyboard control "t" for that.

**Reply by gbishop** [reply]
Maybe the UI didn't have focus and the designer was open? I need to figure out how to force that. Please try clicking with the mouse in the UI or Monitor area to see if that helps.

**Reply by aetkie** [reply]
Clicking the mouse inside designer will allow the keyboard controls to work. But not if an element within designer is selected or highlighted. For instance, I click on the Actions tab but an origin box was highlighted and that obviously did not allow a keyboard shortcut, it only added that letter ("T") to the text.

Clicking the mouse inside the UI did not allow the keyboard controls to work for me.

**Comment by lageist** [comment]
So cool that within seconds of setting it up it's tracking my head movements. Like Andrea notes above, I'm not able to use the keyboard controls to access the tracky-mouse controls to see the camera and play around with the parameters.
